File: archive/python/string/TextJustification.py
class Solution:
    # @param words, a list of strings
    # @param L, an integer
    # @return a list of strings
    def fullJustify(self, words, L):
        result = []
        curr_line = []
        curr_length = 0
        curr = ""
        for i in xrange(len(words)):
            if len(curr_line) == 0 or (curr_length + len(words[i]) + 1) <= L:
                curr_line.append(words[i])
                curr_length += len(words[i])
                if len(curr_line) != 1:
                    curr_length += 1
            else:
                if (len(curr_line) == 1):
                    curr += curr_line[0] + ((L - len(curr_line[0])) * " ")
                else:
                    total_space = L - curr_length
                    per_space = total_space / (len(curr_line) - 1)
                    mod_space = total_space % (len(curr_line) - 1)
                    for j in xrange(len(curr_line)):
                        curr += curr_line[j]
                        if j != len(curr_line) - 1:
                            curr += (per_space + 1) * " ";
                        if mod_space != 0:
                            curr += " "
                            mod_space -= 1
                result.append(curr)
                curr = ""
                curr_line = [words[i]]
                curr_length = len(words[i])

        last_line = ""
        for i in xrange(len(curr_line)):
            if i != 0:
                last_line += " "
            last_line += curr_line[i]

        result.append(last_line + (L - len(last_line)) * " ")

        return result

if __name__ == "__main__":
    solution = Solution()

    print solution.fullJustify(["What", "be"], 12)

File: archive/cpp/string/TextJustification.cpp
#include <vector>
#include <string>

class Solution {
public:
    std::vector<std::string> fullJustify(std::vector<std::string> &words, int L) {
        std::vector<std::string> result;
        std::vector<std::string> curr_line;
        int curr_length = 0;
        std::string curr = "";

        for (int i = 0; i < words.size(); i ++) {
            if (curr_line.size() == 0 || (curr_length + words[i].length() + 1) <= L) {
                curr_line.push_back(words[i]);
                curr_length += words[i].length();
                if (curr_line.size() != 1) {
                    curr_length ++;
                }
            } else {
                if (curr_line.size() == 1) {
                    curr += curr_line[0]  + std::string(L-curr_line[0].length(), ' ');
                } else {
                    int total_space = L - curr_length;
                    int per_space = total_space / (curr_line.size() - 1);
                    int mod_space = total_space % (curr_line.size() - 1);
                    for (int j = 0; j < curr_line.size(); j ++) {
                        curr += curr_line[j];
                        if (j != curr_line.size() - 1) {
                            curr += std::string(per_space+1, ' ');
                        }
                        if (mod_space != 0) {
                            curr += " ";
                            mod_space --;
                        }
                    }
                }
                result.push_back(curr);
                curr = "";
                std::string temp[] = {words[i]};
                curr_line = std::vector<std::string> (temp, temp + sizeof(temp) / sizeof(temp[0]));
                curr_length = words[i].length();
            }
        }
        std::string last_line = "";
        for (int i = 0; i < curr_line.size(); i ++) {
            if (i != 0) {
                last_line += " ";
            }
            last_line += curr_line[i];
        }

        result.push_back(last_line + std::string(L-last_line.length(), ' '));
        return result;
    }
};
